### ISSUE TITLE
Switch Expression (on conditions)

### ISSUE THREAD

**Issue by TonyValenti** [dotnet/csharplang#3350]
It would be great if I could do the following:
````
...
var Value = switch {
   Value1 == Value2 => "Hello",
   Value2 > Value3 => "World",
   Value4 > Value5 && Value6.Foo() => "Complex",
  _ => "Default",
};

````

**Comment by HaloFour** [comment]
https://github.com/dotnet/csharplang/issues/812

**Comment by 333fred** [comment]
I don't think that's quite it, as I'm assuming Tony is looking for non-constant patterns as well? Relational operators are coming in 9.

**Comment by Sunnie-Shine** [comment]
You can use `true switch` with discards.
```csharp
var Value = true switch {
    _ when Value1 == Value2 => "Hello",
    _ when Value2 > Value3 => "World",
    _ when Value4 > Value5 && Value6.Foo() => "Complex",
    _ => "Default",
};
```
Although `true switch` is really used rarely...
See [this link](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpHoAOATgJYBuAhgC4Cm6FAdnegGo0A2Arg3AA0nHvxjCufBqgmiGAFllSArEv4A2ANyFiOomkzz0AWQAUASnx7iRWmRFT0AXnR0y/dAGcA7hToBjAAsrJBsw4gB9dG9AhhYHfjhnF0kxZwA+dAAiAAkGbm4AeyzBa3CiKJi4hIYYdEzU6QzsgHVCsm5gErLyytj4xqMGuWV0ADIxmvUAOgAxQsKLZqzsQoBbEm4GAA9u0PKbKKdMrIARBgAzGl5uOj2DgF9tfaIH5DekZFZGMhYeTDg6kwdQAWiEbAB6CHoADyLG4AE90Bd2ugaK4GJ46NNrAYoID0AAjBbcdDzRZ0QIUTzMNiYKCU6mWY6udwMZ4PIA=).

**Comment by TonyValenti** [comment]
@Sunnie-Shine Wow!  That's sweet!  I'm glad that exists.  It would be  really nice if I could just emit the ````true```` and  ````_ when````.

**Comment by bramborman** [comment]
And what about the ternary operator? The syntax is a bit different but even less verbose :) 

```csharp
...
var Value =
   Value1 == Value2 ? "Hello"
   : Value2 > Value3 ? "World"
   : Value4 > Value5 && Value6.Foo() ? "Complex"
   : "Default"
;
```

**Comment by surfsky** [comment]
The current Pattern Matching switch expression is unreadable,  un-c-like syntax. And it brings 'and' 'or' keywords that  will obscure with '&&' '||' symbols. 

The tranditional c/cpp/c# are syntax-strict, and use () {}  || && to make codes readable and maintainable. Too free and oral syntax will make language unmaintainable,  the pattern matching switch expression is a dangerous start.

Why not change the switch code to c-like syntax:

`
    int age = 10;
    var stage = switch(age)
    {
        (age < 0) =>  LiftStage.Prenatal,
        (age < 2) =>  LifeStage.Infant,
        (age < 4) =>  LifeStage.Toddler,
        _               =>  LifeStage.LateAdult,
    };
`

or

`
    int longvariant = 10;
    var stage = switch(longvariant as t)
    {
        (t < 0) =>  LiftStage.Prenatal,
        (t < 2) =>  LifeStage.Infant,
        (t < 4) =>  LifeStage.Toddler,
        _          =>  LifeStage.LateAdult,
    };
`

or use other keyword "map" to avoid syntax collision?

`
    int t = 10;
    var stage = map(t)
    {
        (t < 0) =>  LiftStage.Prenatal,
        (t < 2) =>  LifeStage.Infant,
        (t < 4) =>  LifeStage.Toddler,
        _          =>  LifeStage.LateAdult,
    };
`

**Comment by HaloFour** [comment]
@surfsky 

The `switch` expression syntax has already shipped as a part of C# 8.0.

And for the `and`/`or` patterns, they were chosen specifically to differ from `&&` and `||` because these patterns can be used in `if` statements where it would be impossible for the compiler to differentiate between them as patterns or as operators.

**Comment by surfsky** [comment]
I maybe know the reason, but this syntax is so ..... oral and unreadable and make cshap much more complex and chaos, It's maybe not a good start.
